pkg/cover/canonicalizer.go

```diff
@@ -0,0 +1,109 @@
+	// Map of modules stored as module name:kernel offset.
+	modules map[string]uint32
+
+	// Contains a sorted list of the canonical module addresses.
+	moduleKeys []uint32
+}
+
+type CanonicalizerInstance struct {
+	canonical Canonicalizer
+
+	// Contains a sorted list of the instance's module addresses.
+	moduleKeys []uint32
+
+	// Contains a map of the uint32 address to the necessary offset.
+	instToCanonicalMap map[uint32]int
+	canonicalToInstMap map[uint32]int
+}
+
+func NewCanonicalizer(modules []host.KernelModule) *Canonicalizer {
+	// Create a map of canonical module offsets by name.
+	canonicalModules := make(map[string]uint32)
+	for _, module := range modules {
+		canonicalModules[module.Name] = uint32(module.Addr)
+	}
+
+	// Store sorted canonical address keys.
+	canonicalModuleKeys := make([]uint32, len(modules))
+	setModuleKeys(canonicalModuleKeys, modules)
+
+	return &Canonicalizer{
+		modules:    canonicalModules,
+		moduleKeys: canonicalModuleKeys,
+	}
+}
+
+func (can *Canonicalizer) NewInstance(modules []host.KernelModule) *CanonicalizerInstance {
+	// Save sorted list of module offsets.
+	moduleKeys := make([]uint32, len(modules))
+	setModuleKeys(moduleKeys, modules)
+
+	// Create a hash between the "canonical" module addresses and each VM instance.
+	instToCanonicalMap := make(map[uint32]int)
+	canonicalToInstMap := make(map[uint32]int)
+	for _, module := range modules {
+		canonicalAddr := can.modules[module.Name]
+		instAddr := uint32(module.Addr)
+		canonicalToInstMap[canonicalAddr] = int(instAddr) - int(canonicalAddr)
+		instToCanonicalMap[instAddr] = int(canonicalAddr) - int(instAddr)
+	}
+
+	return &CanonicalizerInstance{
+		canonical:          *can,
+		moduleKeys:         moduleKeys,
+		instToCanonicalMap: instToCanonicalMap,
+		canonicalToInstMap: canonicalToInstMap,
+	}
+}
+
+func (ci *CanonicalizerInstance) Canonicalize(cov []uint32) {
+	convertModulePCs(ci.moduleKeys, ci.instToCanonicalMap, cov)
+}
+
+func (ci *CanonicalizerInstance) Decanonicalize(cov []uint32) {
+	convertModulePCs(ci.canonical.moduleKeys, ci.canonicalToInstMap, cov)
+}
+
+func setModuleKeys(moduleKeys []uint32, modules []host.KernelModule) {
+	// Store sorted list of addresses. Used to binary search when converting PCs.
+
+	for idx, module := range modules {
+		// Truncate PCs to uint32, assuming that they fit into 32 bits.
+		// True for x86_64 and arm64 without KASLR.
+		moduleKeys[idx] = uint32(module.Addr)
+	}
+	sort.Slice(moduleKeys, func(i, j int) bool { return moduleKeys[i] < moduleKeys[j] })
+}
+
+func convertModulePCs(moduleKeys []uint32, conversionHash map[uint32]int, cov []uint32) {
+	// Skip conversion if modules are not used.
+	if len(moduleKeys) == 0 {
+		return
+	}
+	for idx, pc := range cov {
+		// Determine which module each pc belongs to.
+		i, _ := sort.Find(len(moduleKeys), func(i int) int {
+			if pc < moduleKeys[i] {
+				return -1
+			}
+			return +1
+		})
+		// If address is in a module (above first module address offset),
```

Comment:
Do kernel PC addresses always stay below all module addresses?

Reply:
Good question, thanks! They did for x86, but I looked into it and the module addresses are below the kernel PCs for the arm64 device I tested. I modified the code to include the module size when parsing `proc/modules` in `pkg/host/machine_info_linux.go` so we can determine the end of the modules to separate them from kernel PC addresses.

An alternative would be to instead find the beginning of the kernel PC addresses from the `vmlinux` file, but I felt it may be safer to assume anything outside of the module space is in the main kernel. I can switch to this method though if you think it would be better not to pass the extra module size argument from the device.

Reply:
I think an even safer approach would be to assume that all modules together do not make a contiguous memory block and to only convert PCs that fall into [module_i_start;module_i_end] ranges. All other PCs would remain unaffected.

It shouldn't affect performance -- you can still do binary search e.g. on the module start PC and just check whether the PC-to-be converted belongs to the range found by the binary search.

Reply:
That's a really good point. I've updated it to include the module end for all modules, and added values in between/above modules in the test code.